File: basic_dsa_one_hundred_three.java
/**
 * Basic DSA Problems - First Bad Version
 * Find the first bad version using binary search (minimize API calls)
 */
public class basic_dsa_one_hundred_three {

    // Problem: You have n versions [1, 2, ..., n]. isBadVersion(version) returns true if version is bad.
    // All versions after a bad one are bad. Find the first bad version with minimum API calls.

    // Simulated bad version threshold for testing
    private static int badFrom = 4;

    public static void setBadFrom(int version) {
        badFrom = version;
    }

    static boolean isBadVersion(int version) {
        return version >= badFrom;
    }

    /**
     * Binary search: if mid is bad, first bad is mid or left; else search right.
     * Time O(log n), Space O(1)
     */
    public static int firstBadVersion(int n) {
        int left = 1, right = n;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (isBadVersion(mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: First Bad Version");
        System.out.println("=".repeat(60));

        setBadFrom(4);
        System.out.println("n = 5, first bad = 4 -> " + firstBadVersion(5));

        setBadFrom(1);
        System.out.println("n = 1, first bad = 1 -> " + firstBadVersion(1));

        setBadFrom(2);
        System.out.println("n = 2, first bad = 2 -> " + firstBadVersion(2));
    }
}

File: basic_dsa_one_hundred_three.py
"""
Basic DSA - Lowest Common Ancestor of a Binary Search Tree
Given a BST, find the lowest common ancestor (LCA) of two nodes.
"""


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def lowest_common_ancestor(root, p, q):
    """
    Use BST property: move left/right until split point.
    Time Complexity: O(h)
    Space Complexity: O(1)
    """
    if p.val > q.val:
        p, q = q, p

    cur = root
    while cur:
        if q.val < cur.val:
            cur = cur.left
        elif p.val > cur.val:
            cur = cur.right
        else:
            return cur
    return None


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Lowest Common Ancestor of a BST")
    print("=" * 60)

    #        6
    #      /   \
    #     2     8
    #    / \   / \
    #   0   4 7   9
    #      / \
    #     3   5
    n0 = TreeNode(0)
    n3 = TreeNode(3)
    n5 = TreeNode(5)
    n4 = TreeNode(4, n3, n5)
    n2 = TreeNode(2, n0, n4)
    n7 = TreeNode(7)
    n9 = TreeNode(9)
    n8 = TreeNode(8, n7, n9)
    root = TreeNode(6, n2, n8)

    print(lowest_common_ancestor(root, n2, n8).val)  # 6
    print(lowest_common_ancestor(root, n2, n4).val)  # 2

